src/core/control/actions/ActionDatabase.cpp

```diff
@@ -0,0 +1,166 @@
+                              << "\" = " << std::setw(7) << ActionProperties<a>::initialState(db->control)
+                              << " |            |");
+
+        db->gActions[a].reset(g_simple_action_new_stateful(Action_toString(a), nullptr,
+                                                           makeGVariant<typename ActionProperties<a>::state_type>(
+                                                                   ActionProperties<a>::initialState(db->control))),
+                              xoj::util::adopt);
+
+        finishSetup<a>(db, "change-state");
+    }
+
+    // Actions with both state and parameter (with matching type)
+    template <Action a, std::enable_if_t<has_param<a>() && has_state<a>(), bool> = true>
+    static void assign(ActionDatabase* db) {
+        static_assert(
+                std::is_same_v<typename ActionProperties<a>::state_type, typename ActionProperties<a>::parameter_type>);
+
+        PRINT_DEBUG(START_ROW << " | \"" << (const char*)gVariantType<typename ActionProperties<a>::state_type>()
+                              << "\" = " << std::setw(7) << ActionProperties<a>::initialState(db->control)
+                              << " | type = \"" << (const char*)gVariantType<typename ActionProperties<a>::state_type>()
+                              << "\" |");
+
+        db->gActions[a].reset(g_simple_action_new_stateful(Action_toString(a),
+                                                           gVariantType<typename ActionProperties<a>::state_type>(),
+                                                           makeGVariant<typename ActionProperties<a>::state_type>(
+                                                                   ActionProperties<a>::initialState(db->control))),
+                              xoj::util::adopt);
+
+        finishSetup<a>(db, "change-state");
+    }
+
+    template <size_t... As>
+    static void populateImpl(std::index_sequence<As...>, ActionDatabase* db) {
+        ((assign<static_cast<Action>(As)>(db)), ...);
+    }
+
+
+public:
+    static void populate(ActionDatabase* db);
+};
+
+void ActionDatabase::Populator::populate(ActionDatabase* db) {
+    PRINT_DEBUG("Populating ActionDatabase:");
+    PRINT_DEBUG("        ACTION NAME:                |  STATE INIT   | PARAM TYPE |");
+
+    populateImpl(std::make_index_sequence<static_cast<size_t>(Action::_MAX_VALUE) + 1>(), db);
+}
+
+ActionDatabase::ActionDatabase(Control* control):
+        control(control), win(GTK_APPLICATION_WINDOW(control->getWindow()->getWindow())) {
+    Populator::populate(this);
+}
+
+ActionDatabase::~ActionDatabase() {
+    auto sig = signalIds.begin();
+    for (auto& a: gActions) {
+        /**
+         * The GAction's might still be referenced elsewhere (e.g. by the GtkApplicationWindow), so we need to
+         * disconnect the signals in case a callback is called and the Control instance has been destroyed
+         */
+        g_signal_handler_disconnect(a.get(), *sig++);
```

Comment:
**Question:** This only works, if the action Database is killed first, right? In my opinion, all signals must be canceled/disconnected by the owner of the referenced captures

Reply:
Yes, but the Control instance owns the ActionDatabase instance, so the ActionDatabase instance will get destroyed first in every situation.